onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -97,64 +91,75 @@ inline void CastToStringData(const Tensor* in, Tensor* out, const TensorShape& s
   }
 }
 
+template <typename SrcType>
+typename std::enable_if<!std::is_floating_point<SrcType>::value, void>::type
+CastToStringData(const Tensor& in, Tensor& out, const TensorShape& shape) {
+  const int64_t len = shape.Size();
+  const auto input_data = in.DataAsSpan<SrcType>();
+  auto output_data = out.MutableDataAsSpan<std::string>();
+
+  for (int i = 0; i < len; ++i) {
+    output_data[i] = (std::ostringstream() << input_data[i]).str();
+  }
+}
+
 template <typename DstType>
-inline void CastFromStringData(const Tensor* in, Tensor* out, const TensorShape& shape) {
-  if (std::is_same<DstType, std::string>::value) return;
+void CastFromStringData(const Tensor& in, Tensor& out, const TensorShape& shape) {
   const int64_t len = shape.Size();
   ORT_ENFORCE(len > 0);
   if (std::is_same<DstType, float>::value) {
-    auto* mutable_data = out->MutableData<float>();
+    auto* mutable_data = out.MutableData<float>();
     for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stof(in->Data<std::string>()[i]);
+      mutable_data[i] = std::stof(in.Data<std::string>()[i]);
```

Comment:
not related to the changes in this PR, but do we want to accept stuff like "  1.2abc" as 1.2? i think these std::sto[x] calls will process as many characters as possible from the string, and possibly not all of them.

Reply:
Whilst it would be more correct, I don't think it's worth the cost to validate the characters in the strings. Considering that each input value is in a separate string, any strange character will only affect the current value and not subsequent ones.

---
In reply to: [469626219](https://github.com/microsoft/onnxruntime/pull/4765#discussion_r469626219) [](ancestors = 469626219)